src/master-playlist-controller.js

```diff
@@ -831,40 +847,81 @@ export default class MasterPlaylistController extends videojs.EventTarget {
     });
   }
 
-  updateCues_(media) {
+  updateCues_(media, offset = 0) {
     if (!this.useCueTags_ || !media.segments) {
       return;
     }
 
-    while (this.cueTagsTrack_.cues.length) {
-      this.cueTagsTrack_.removeCue(this.cueTagsTrack_.cues[0]);
-    }
-
-    let mediaTime = 0;
+    let mediaTime = offset;
+    let track = this.cueTagsTrack_;
+    let cue;
 
     for (let i = 0; i < media.segments.length; i++) {
       let segment = media.segments[i];
 
-      if ('cueOut' in segment || 'cueOutCont' in segment || 'cueIn' in segment) {
-        let cueJson = {};
+      if (cue !== undefined) {
 
-        if ('cueOut' in segment) {
-          cueJson.cueOut = segment.cueOut;
+        if ('cueIn' in segment) {
+          // Found a CUE-IN so end the cue
+          if (mediaTime !== cue.endTime) {
+            // If the end time does not match the current mediaTime, a CUE-IN may have
+            // been inserted earlier than expected, so the cue should be updated.
+            cue.endTime = mediaTime;
+            cue.adEndTime = mediaTime;
+          }
+          mediaTime += segment.duration;
+          cue = undefined;
+          continue;
         }
-        if ('cueOutCont' in segment) {
-          cueJson.cueOutCont = segment.cueOutCont;
+
+        if (mediaTime < cue.endTime) {
+          // Already processed this mediaTime for this cue
```

Comment:
It may just be me, but the wording here threw me off a bit. Are you saying that we've already adjusted the current cue's end time to account for this segment's duration on a prior call to updateCues_?